go-client/client/app_stub.go

```diff
@@ -0,0 +1,122 @@
+	cancelAck <-chan struct{}
+}
+
+func newAppStub(client *Client) appStub {
+	return appStub{client: client, stub: apppb2.NewApplicationServiceClient(client.grpcChannel)}
+}
+
+type changeHandler func(update *apppb2.FieldsChangeUpdate)
+
+// Note that the provided context is saved and used to continue reading the stream,
+// so any timeouts or cancellations that have been set for it will affect the listFields stream.
+// A thread that continues reading the listFields request and updating the client's tables will start in the background
+// if FetchRepeating is specified.
+// The client lock should be held when calling this function.
+func (as *appStub) listFields(ctx context.Context, fetchOption FetchOption, handler changeHandler) error {
+	ctx, err := as.client.withToken(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	as.cancelFetchLoop()
+
+	req := apppb2.ListFieldsRequest{}
+	fieldStream, err := as.stub.ListFields(ctx, &req)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	resp, err := fieldStream.Recv()
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	handler(resp)
+
+	as.cancelFunc = cancel
+
+	if fetchOption == FetchRepeating {
+		cancelAck := make(chan struct{})
+		as.cancelAck = cancelAck
+		go fetchTablesLoop(fieldStream, handler, cancelAck)
+		return nil
+	} else {
+		as.cancelFetchLoop()
+		return nil
+	}
+}
+
+// isFetching returns true if an existing FetchTables loop is running.
+// The client lock should be held when calling this function.
+func (as *appStub) isFetching() bool {
+	return as.cancelFunc != nil
+}
+
+// cancelFetchLoop cancels an existing FetchTables loop, if one exists.
+// If one exists, this will block until the loop has finished writing any results that are in-progress.
+// The client lock should be held when calling this function.
+func (as *appStub) cancelFetchLoop() {
+	if as.cancelFunc != nil {
+		as.cancelFunc()
+		as.cancelFunc = nil
+	}
+
+	if as.cancelAck != nil {
+		<-as.cancelAck
+		as.cancelAck = nil
+	}
+}
+
+// Close closes the app stub and frees any associated resources.
+// The app stub should not be used after calling this function.
+// The client lock should be held when calling this function.
+func (as *appStub) Close() {
+	as.cancelFetchLoop()
+}
+
+// fetchTablesLoop runs a loop that will check for new changes to fields.
+// When a new change occurs, it will pass the change to the given handler.
+// When the context for the stream gets canceled, the loop will halt
+// and close the cancelAck channel to signal that it is finished.
+func fetchTablesLoop(stream apppb2.ApplicationService_ListFieldsClient, handler changeHandler, cancelAck chan<- struct{}) {
+	for {
+		resp, err := stream.Recv()
+		if err != nil {
+			if status, ok := status.FromError(err); ok && status.Code() == codes.Canceled {
+				close(cancelAck)
+				break
+			}
+			fmt.Println("failed to list fields: ", err)
+			return
+		}
+
+		handler(resp)
+	}
+}
```

Comment:
I have concerns that the context model for all of the table fetching is jacked up.  In my other comment, it looks like we are switching contexts.  This needs to be thought through very carefully.